lua/lspkind/init.lua

```diff
@@ -1,185 +1,309 @@
-
-  local symbol_map = kind_presets[preset]
-  lspkind.symbol_map = (opts and opts['symbol_map'] and
-                       vim.tbl_extend('force', symbol_map, opts['symbol_map'])) or symbol_map
+-- Setup plugin
+function M.setup(opts)
+  opts = vim.tbl_deep_extend('force', H.defaults, opts or {})
 
-  local symbols = {}
-  local len = kind_len
-  for i = 1, len do
-    local name = kind_order[i]
-    symbols[i] = lspkind.symbolic(name, opts)
+  -- Store symbols, globally accessible
+  if type(opts.symbols) == 'string' then
+    M.symbols = kinds.presets[opts.symbols]
+  else
+    M.symbols = vim.tbl_deep_extend('force', kinds.presets.default, opts.symbols)
   end
 
-  for k,v in pairs(symbols) do
-    require('vim.lsp.protocol').CompletionItemKind[k] = v
-  end
+  H.set_symbols(kinds.list.cmp, kinds.order.cmp, M.symbols, opts.mode)
+  H.set_symbols(kinds.list.sym, kinds.order.sym, M.symbols, opts.mode)
 end
 
-lspkind.presets = kind_presets
-lspkind.symbol_map = kind_presets.default
-
-function lspkind.symbolic(kind, opts)
-  local mode = opt_mode(opts)
-  local formatter = modes[mode]
+-- nvim-cmp
+function M.cmp_format(opts)
+  opts = vim.tbl_deep_extend('force', H.defaults, opts or {})
 
-  -- if someone enters an invalid mode, default to symbol
-  if formatter == nil then
-      formatter = modes['symbol']
-  end
-
-  return formatter(kind)
-end
-
-function lspkind.cmp_format(opts)
-  if opts == nil then
-    opts = {}
-  end
-  if opts.preset or opts.symbol_map then
-    lspkind.init(opts)
+  if opts.symbols then
+    M.setup(opts)
   end
 
   return function(entry, vim_item)
     if opts.before then
-          vim_item = opts.before(entry, vim_item)
+      vim_item = opts.before(entry, vim_item)
     end
-        
-    vim_item.kind = lspkind.symbolic(vim_item.kind, opts)
+
+    vim_item.kind = H.get_symbol(kinds.list.cmp, kinds.order.cmp, vim_item.kind).icon
 
     if opts.menu ~= nil then
       vim_item.menu = opts.menu[entry.source.name]
     end
 
     if opts.maxwidth ~= nil then
-        vim_item.abbr = string.sub(vim_item.abbr, 1, opts.maxwidth)
+      vim_item.abbr = string.sub(vim_item.abbr, 1, opts.maxwidth)
     end
 
     return vim_item
   end
 end
 
-return lspkind
+-- Deprecation warnings!
+function M.init(opts)
+  local warns = {}
+  table.insert(warns, 'lspkind.init() was replaced by lspkind.setup()\nPlease update your settings!')
+  if opts ~= nil then
+    if opts['with_text'] ~= nil then
+      table.insert(warns, 'setup({ with_text }) was replaced by setup({ mode }).\nPlease update your settings!')
+    end
+    if opts.symbol_map ~= nil then
+      table.insert(warns, 'setup({ symbol_map }) was replaced by setup({ symbols })\nPlease update your settings!')
+      table.insert(warns, 'setup({ preset }) was replaced by setup({ symbols })\nPlease update your settings!')
+      table.insert(warns, 'lspkind.symbol_map was replaced by lspkind.symbols.\nPlease update your settings!')
+    end
```

Comment:
Why are you trying to deprecate this? Can't we just keep the existing config? What's wrong with them?

Reply:
Deprecating these options simplifies configuration
As I had explained before, if setup({ symbols }) is a table the defined symbols will be used, if it is string the corresponding pattern will be used